pkg/controller/user-cluster-controller-manager/ccm-csi-migrator/controller.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ccmcsimigrator
+
+import (
+	"context"
+	"fmt"
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/common"
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+	"go.uber.org/zap"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1/helper"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"strconv"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ccm-csi-migrator"
+)
+
+type reconciler struct {
+	log          *zap.SugaredLogger
+	seedClient   ctrlruntimeclient.Client
+	userClient   ctrlruntimeclient.Client
+	seedRecorder record.EventRecorder
+	versions     kubermatic.Versions
+	clusterName  string
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, seedMgr, userMgr manager.Manager, versions kubermatic.Versions, clusterName string) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		log:          log,
+		seedClient:   seedMgr.GetClient(),
+		userClient:   userMgr.GetClient(),
+		seedRecorder: seedMgr.GetEventRecorderFor(controllerName),
+		versions:     versions,
+		clusterName:  clusterName,
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	commonHandler := func(obj ctrlruntimeclient.Object, queue workqueue.RateLimitingInterface) {
```

Comment:
In this case, using a simple helper like this one wouldn't fit because we need to check the annotations on the cluster resource, and not on the triggering one (the machine). BTW, I removed the filter because the ccm-migration flag is injected by the seed-controller-manager only when the migration is strictly required (this condition is checked a priori).